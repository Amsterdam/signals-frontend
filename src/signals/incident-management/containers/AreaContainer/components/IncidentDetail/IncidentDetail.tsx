import {
  Heading,
  themeSpacing,
  Link as AscLink,
  themeColor,
} from '@amsterdam/asc-ui'
import { Link } from 'react-router-dom'
import styled from 'styled-components'

import BackLink from 'components/BackLink'
import { string2date, string2time } from 'shared/services/string-parser'
import { isStatusEnd } from 'signals/incident-management/definitions/statusList'
import type { Incident } from 'types/api/incident'
import type { StatusCode } from 'types/status-code'

interface IncidentDetailProps {
  incident: Incident
  onBack: () => void
}

const Wrapper = styled.section`
  padding: ${themeSpacing(4)};
`

const StyledBackLink = styled(BackLink)`
  margin-bottom: ${themeSpacing(4)};
`

const StyledLink = styled(AscLink)`
  margin-bottom: ${themeSpacing(4)};
  text-decoration: underline;
  display: block;

  :hover {
    cursor: pointer;
    & > * {
      color: ${themeColor('secondary')};
    }
  }
`

const DefinitionList = styled.div`
  margin-top: ${themeSpacing(4)};
`

const SectionTerm = styled.dt`
  color: ${themeColor('tint', 'level5')};
  margin-bottom: ${themeSpacing(1)};
`

const SectionDescription = styled.dd`
  margin-bottom: ${themeSpacing(4)};
`

const Status = styled.div<{ isEnded: boolean }>`
  font-weight: 700;
  color: ${({ isEnded }) =>
    themeColor('support', isEnded ? 'valid' : 'invalid')};
`

const StyledParagraph = styled.p`
  font-weight: 700;
`

const formatDate = (date: string): string => {
  return `${string2date(date)} ${string2time(date)}`
}

const getIncidentTitlePrefix = (incident: Incident) => {
  const parentId = incident._links?.['sia:parent']?.href?.split('/').pop()
  const hasChildren = incident._links?.['sia:children']?.length

  if (hasChildren) {
    return 'Hoofd'
  }

  if (parentId) {
    return 'Deel'
  }

  return 'Standaard'
}

const IncidentDetail: React.FC<IncidentDetailProps> = ({
  onBack,
  incident,
}) => {
  return (
    <Wrapper>
      <StyledBackLink to="#" onClick={() => onBack()}>
        Terug naar filter
      </StyledBackLink>

      <StyledLink
        forwardedAs={Link}
        to={`/manage/incident/${incident.id}`}
        target="_blank"
      >
        <StyledParagraph data-testid="incident-heading">
          {`${getIncidentTitlePrefix(incident)}melding ${incident.id}`}
        </StyledParagraph>
      </StyledLink>

      <Heading as="h2" styleAs="h4" data-testid="text">
        {incident.text}
      </Heading>

      <DefinitionList>
        <SectionTerm data-testid="location-label">Locatie</SectionTerm>
        <SectionDescription>
          <span data-testid="location">
            {incident.location?.address_text || 'Locatie is gepind op de kaart'}
          </span>
        </SectionDescription>

        <SectionTerm data-testid="date-label">Gemeld op</SectionTerm>
        <SectionDescription>
          <span data-testid="date">
            {incident.created_at && formatDate(incident.created_at)}
          </span>
        </SectionDescription>

        <SectionTerm data-testid="status-label">Status</SectionTerm>
        {incident.status && (
          <SectionDescription>
            <Status isEnded={isStatusEnd(incident.status.state as StatusCode)}>
              <span data-testid="status">{incident.status.state_display}</span>
            </Status>
          </SectionDescription>
        )}

        <SectionTerm data-testid="subcategory-label">
          Subcategorie (verantwoordelijke afdeling)
        </SectionTerm>
        <SectionDescription>
          <span data-testid="subcategory">{incident.category?.sub} </span>
          <span data-testid="departments">
            ({incident.category?.departments})
          </span>
        </SectionDescription>
      </DefinitionList>
    </Wrapper>
  )
}

export default IncidentDetail
