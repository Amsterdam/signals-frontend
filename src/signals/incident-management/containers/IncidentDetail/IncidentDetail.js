// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2021 Gemeente Amsterdam
import { useReducer, useEffect, useCallback, useState } from 'react'
import { useParams } from 'react-router-dom'
import { themeSpacing, Row, Column } from '@amsterdam/asc-ui'
import styled from 'styled-components'
import { useDispatch, useSelector } from 'react-redux'

import configuration from 'shared/services/configuration/configuration'
import { makeSelectSubCategories } from 'models/categories/selectors'
import { useFetch, useEventEmitter, useFetchAll } from 'hooks'
import { showGlobalNotification } from 'containers/App/actions'
import { VARIANT_ERROR, TYPE_LOCAL } from 'containers/Notification/constants'
import { getErrorMessage } from 'shared/services/api/api'
import { patchIncidentSuccess } from 'signals/incident-management/actions'
import History from 'components/History'
import reducer, { initialState } from './reducer'

import Attachments from './components/Attachments'
import ChildIncidents from './components/ChildIncidents'
import DetailHeader from './components/DetailHeader'
import MetaList from './components/MetaList'
import LocationForm from './components/LocationForm'
import AttachmentViewer from './components/AttachmentViewer'
import Detail from './components/Detail'
import LocationPreview from './components/LocationPreview'
import CloseButton from './components/CloseButton'
import IncidentDetailContext from './context'

import {
  CLOSE_ALL,
  EDIT,
  PATCH_START,
  PATCH_SUCCESS,
  PREVIEW,
  RESET,
  SET_ATTACHMENTS,
  SET_CHILDREN,
  SET_CHILDREN_HISTORY,
  SET_CHILD_INCIDENTS,
  SET_CONTEXT,
  SET_DEFAULT_TEXTS,
  SET_ERROR,
  SET_HISTORY,
  SET_INCIDENT,
} from './constants'
import useUpload from './hooks/useUpload'

const StyledRow = styled(Row)`
  position: relative;
`

const StyledAttachments = styled(Attachments)`
  margin-bottom: ${themeSpacing(4)};
`

const DetailContainer = styled(Column)`
  flex-direction: column;
  position: relative;
  z-index: 1;
  justify-content: flex-start;
`

const Preview = styled.div`
  background: white;
  bottom: 0;
  height: 100%;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
  width: 100%;
  z-index: 1;
`

const IncidentDetail = () => {
  const { emit, listenFor, unlisten } = useEventEmitter()
  const storeDispatch = useDispatch()
  const { id } = useParams()
  const [isRemovingAttachment, setRemovingAttachment] = useState(false)
  const [isParent, setIsParent] = useState(false)
  const [isChild, setIsChild] = useState(false)
  const [state, dispatch] = useReducer(reducer, initialState)
  const {
    error,
    get: getIncident,
    data: incident,
    isSuccess,
    patch,
  } = useFetch()
  const { get: getHistory, data: history } = useFetch()
  const {
    get: getAttachments,
    data: attachments,
    isLoading: isAttachmentsLoading,
  } = useFetch()
  const {
    del: deleteAttachment,
    isLoading: isDeleteAttachmentLoading,
    isSuccess: isDeleteAttachmentSuccess,
  } = useFetch()
  const { get: getDefaultTexts, data: defaultTexts } = useFetch()
  const { get: getChildren, data: children } = useFetch()
  const { get: getChildrenHistory, data: childrenHistory } = useFetchAll()
  const { get: getContext, data: context } = useFetch()
  const { get: getChildIncidents, data: childIncidents } = useFetchAll()
  const { upload, uploadSuccess, uploadProgress, uploadError } = useUpload()

  const subcategories = useSelector(makeSelectSubCategories)

  useEffect(() => {
    document.addEventListener('keyup', handleKeyUp)

    return () => {
      document.removeEventListener('keyup', handleKeyUp)
    }
  }, [handleKeyUp, listenFor, unlisten])

  useEffect(() => {
    dispatch({ type: SET_ERROR, payload: error })

    if (error) {
      const title =
        error.status === 401 || error.status === 403
          ? 'Geen bevoegdheid'
          : 'Bewerking niet mogelijk'
      const message = getErrorMessage(
        error,
        'Deze wijziging is niet toegestaan in deze situatie.'
      )

      storeDispatch(
        showGlobalNotification({
          title,
          message,
          variant: VARIANT_ERROR,
          type: TYPE_LOCAL,
        })
      )
    }
  }, [error, storeDispatch])

  useEffect(() => {
    if (!history) return

    dispatch({ type: SET_HISTORY, payload: history })
  }, [history])

  useEffect(() => {
    if (!attachments) return

    dispatch({ type: SET_ATTACHMENTS, payload: attachments.results })
  }, [attachments])

  useEffect(() => {
    if (!incident?.category) return

    const { main_slug, sub_slug } = incident.category
    getDefaultTexts(
      `${configuration.TERMS_ENDPOINT}${main_slug}/sub_categories/${sub_slug}/status-message-templates`
    )
  }, [incident?.category, getDefaultTexts])

  useEffect(() => {
    if (!defaultTexts) return

    dispatch({ type: SET_DEFAULT_TEXTS, payload: defaultTexts })
  }, [defaultTexts])

  useEffect(() => {
    if (!children) return

    dispatch({ type: SET_CHILDREN, payload: children })
  }, [children])

  useEffect(() => {
    if (!childrenHistory) return

    dispatch({ type: SET_CHILDREN_HISTORY, payload: childrenHistory })
  }, [childrenHistory])

  useEffect(() => {
    if (!childIncidents) return

    dispatch({ type: SET_CHILD_INCIDENTS, payload: childIncidents })
  }, [childIncidents])

  useEffect(() => {
    if (!context) return

    dispatch({ type: SET_CONTEXT, payload: context })
  }, [context])

  useEffect(() => {
    if (!id) return

    dispatch({ type: RESET })
    getIncident(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}`)
  }, [getIncident, id])

  useEffect(() => {
    if (!isSuccess || !state.patching) return

    emit('highlight', { type: state.patching })
    dispatch({ type: PATCH_SUCCESS, payload: state.patching })
    storeDispatch(patchIncidentSuccess())
  }, [isSuccess, state.patching, emit, storeDispatch])

  useEffect(() => {
    if (!incident) return
    dispatch({ type: SET_INCIDENT, payload: incident })

    retrieveUnderlyingData()
  }, [incident, retrieveUnderlyingData])

  const retrieveUnderlyingData = useCallback(() => {
    getHistory(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/history`)

    if (!state.attachments) {
      getAttachments(
        `${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/attachments`
      )
    }

    const isParent = incident._links['sia:children']?.length > 0
    setIsParent(isParent)
    const isChild = Boolean(incident._links['sia:parent'])
    setIsChild(isChild)

    if (isParent) {
      getChildren(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/children/`)
    }

    getContext(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/context`)
  }, [
    getHistory,
    id,
    state.attachments,
    incident?._links,
    getAttachments,
    getChildren,
    getContext,
  ])

  useEffect(() => {
    if (children?.results.length > 0) {
      const viewableResults = children.results.filter(
        (result) => result.can_view_signal
      )

      const childIncidentUrls = viewableResults.map(
        ({ id: childId }) =>
          `${configuration.INCIDENT_PRIVATE_ENDPOINT}${childId}`
      )
      const childrenHistoryUrls = viewableResults.map(
        ({ id: childId }) =>
          `${configuration.INCIDENT_PRIVATE_ENDPOINT}${childId}/history`
      )

      getChildrenHistory(childrenHistoryUrls)
      getChildIncidents(childIncidentUrls)
    }
  }, [children, getChildrenHistory, getChildIncidents])

  const handleKeyUp = useCallback(
    (event) => {
      switch (event.key) {
        case 'Esc':
        case 'Escape':
          closeDispatch()
          break

        default:
          break
      }
    },
    [closeDispatch]
  )

  const updateDispatch = useCallback(
    (action) => {
      dispatch({ type: PATCH_START, payload: action.type })
      patch(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}`, action.patch)
    },
    [id, patch]
  )

  const previewDispatch = useCallback((section, payload) => {
    dispatch({ type: PREVIEW, payload: { preview: section, ...payload } })
  }, [])

  const editDispatch = useCallback((section, payload) => {
    dispatch({ type: EDIT, payload: { edit: section, ...payload } })
  }, [])

  const closeDispatch = useCallback(() => {
    dispatch({ type: CLOSE_ALL })
  }, [])

  const addAttachment = useCallback(
    (files) => {
      if (incident) {
        upload(files, incident.id)
      }
    },
    [incident, upload]
  )

  const removeAttachment = useCallback(
    (attachment) => {
      deleteAttachment(attachment._links.self.href)
    },
    [deleteAttachment]
  )

  useEffect(() => {
    if (
      uploadSuccess ||
      (!isDeleteAttachmentLoading && isDeleteAttachmentSuccess)
    ) {
      getHistory(`${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/history`)
      getAttachments(
        `${configuration.INCIDENT_PRIVATE_ENDPOINT}${id}/attachments`
      )
    }
  }, [
    getAttachments,
    getHistory,
    id,
    uploadSuccess,
    isDeleteAttachmentSuccess,
    isDeleteAttachmentLoading,
  ])

  useEffect(() => {
    if (isAttachmentsLoading || isDeleteAttachmentLoading)
      setRemovingAttachment(true)
  }, [isAttachmentsLoading, isDeleteAttachmentLoading])

  useEffect(() => {
    if (!isAttachmentsLoading) setRemovingAttachment(false)
  }, [isAttachmentsLoading])

  if (!state.incident || !subcategories) return null

  return (
    <IncidentDetailContext.Provider
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      value={{
        incident: state?.incident,
        update: updateDispatch,
        preview: previewDispatch,
        edit: editDispatch,
        close: closeDispatch,
      }}
    >
      <Row data-testid="incidentDetail">
        <Column span={12}>
          <DetailHeader />
        </Column>
      </Row>

      <StyledRow>
        <DetailContainer
          span={{ small: 1, medium: 2, big: 5, large: 7, xLarge: 7 }}
        >
          <Detail context={state.context} />

          <StyledAttachments
            attachments={state.attachments || []}
            add={addAttachment}
            remove={removeAttachment}
            isChildIncident={isChild}
            isParentIncident={isParent}
            isRemoving={isRemovingAttachment}
            uploadProgress={uploadProgress}
            uploadSuccess={uploadSuccess}
            uploadError={uploadError}
          />

          {state.children?.results && state.childrenHistory && (
            <ChildIncidents
              childrenList={state.children.results}
              parent={state.incident}
              history={state.childrenHistory}
              childIncidents={state.childIncidents}
            />
          )}

          {state.history && <History list={state.history} />}
        </DetailContainer>

        <DetailContainer
          span={{ small: 4, medium: 4, big: 4, large: 5, xLarge: 5 }}
          push={{ small: 0, medium: 0, big: 0, large: 0, xLarge: 0 }}
        >
          <MetaList
            defaultTexts={state.defaultTexts}
            childIncidents={state.children?.results}
          />
        </DetailContainer>

        {(state.preview || state.edit) && (
          <Preview>
            {state.preview === 'location' && <LocationPreview />}

            {state.edit === 'location' && <LocationForm />}

            {state.preview === 'attachment' && (
              <AttachmentViewer
                attachments={state.attachments}
                href={state.attachmentHref}
              />
            )}
          </Preview>
        )}
        {state.preview && (
          <CloseButton aria-label="Sluiten" onClick={closeDispatch} />
        )}
      </StyledRow>
    </IncidentDetailContext.Provider>
  )
}

export default IncidentDetail
