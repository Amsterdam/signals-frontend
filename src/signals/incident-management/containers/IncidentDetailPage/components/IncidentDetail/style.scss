@import "config.scss";

.incident-detail {
  &__body {
    &__definition, &__value {
      display: inline-block;
      line-height: 36px;
      vertical-align: top;
    }

    &__definition {
      width: 40%;
      color: $ams-donkergrijs;
      position: relative;
      z-index: 10;
    }

    &__value {
      width: 60%;
      @include avenirnextdemi();
      position: relative;
      z-index: 10;
    }

    &--highlight {
      position: relative;

      &:after {
        content: "";
        position: absolute;
        left: 0;
        top: 0;
        bottom: 0;
        right: 0;
        background-color: $ams-geel;
        animation: highlight-fade-out 2.5s;
      }
    }

    &__link {
      margin-right: 10px;
    }
}

  @keyframes highlight-fade-out {
    from {
      opacity: 1;
    }
    70% {
      opacity: 1;
    }
    to {
      opacity: 0;
    }
  }
}
