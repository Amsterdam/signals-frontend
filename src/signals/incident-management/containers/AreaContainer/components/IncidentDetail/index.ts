import IncidentDetail from './IncidentDetail'

export default IncidentDetail
