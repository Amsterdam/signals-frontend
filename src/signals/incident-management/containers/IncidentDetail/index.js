import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { bindActionCreators } from 'redux';
import isEqual from 'lodash.isequal';
import { Row, Column } from '@datapunt/asc-ui';
import styled from 'styled-components';

import { incidentType, dataListType, defaultTextsType, attachmentsType, historyType } from 'shared/types';

import LoadingIndicator from 'shared/components/LoadingIndicator';
import { makeSelectLoading, makeSelectError } from 'containers/App/selectors';
import {
  requestIncident,
  requestAttachments,
  requestDefaultTexts,
  dismissError,
} from 'models/incident/actions';
import { requestHistoryList } from 'models/history/actions';
import makeSelectIncidentModel from 'models/incident/selectors';
import makeSelectHistoryModel from 'models/history/selectors';
import History from 'components/History';

import './style.scss';

import ChildIncidents from './components/ChildIncidents';
import DetailHeader from './components/DetailHeader';
import MetaList from './components/MetaList';
import AddNote from './components/AddNote';
import LocationForm from './components/LocationForm';
import AttachmentViewer from './components/AttachmentViewer';
import StatusForm from './components/StatusForm';
import Detail from './components/Detail';
import LocationPreview from './components/LocationPreview';

const DetailContainer = styled(Column)`
  flex-direction: column;
  position: relative;
`;

export class IncidentDetail extends React.Component {
  // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);

    this.state = {
      previewState: props.previewState, // showLocation, editLocation, editStatus, showImage
      attachmentHref: props.attachmentHref,
    };

    this.onShowLocation = this.onShowLocation.bind(this);
    this.onEditLocation = this.onEditLocation.bind(this);
    this.onEditStatus = this.onEditStatus.bind(this);
    this.onShowAttachment = this.onShowAttachment.bind(this);
    this.onCloseAll = this.onCloseAll.bind(this);
  }

  componentDidMount() {
    this.fetchAll();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.fetchAll();
    }

    /* istanbul ignore else */
    if (this.props.incidentModel.incident) {
      const category = this.props.incidentModel.incident.category;
      if (
        !isEqual(
          prevProps.incidentModel.incident && prevProps.incidentModel.incident.category,
          this.props.incidentModel.incident.category
        )
      ) {
        this.props.onRequestDefaultTexts({
          main_slug: category.main_slug,
          sub_slug: category.sub_slug,
        });
      }
    }
  }

  onShowLocation() {
    this.setState({
      previewState: 'showLocation',
      attachmentHref: '',
    });
  }

  onEditLocation() {
    this.setState({
      previewState: 'editLocation',
      attachmentHref: '',
    });
  }

  onEditStatus() {
    this.setState({
      previewState: 'editStatus',
      attachmentHref: '',
    });
  }

  onShowAttachment(attachmentHref) {
    this.setState({
      previewState: 'showImage',
      attachmentHref,
    });
  }

  onCloseAll() {
    this.setState({
      previewState: '',
      attachmentHref: '',
    });
  }

  fetchAll() {
    this.props.onRequestIncident(this.props.match.params.id);
    this.props.onRequestHistoryList(this.props.match.params.id);
    this.props.onRequestAttachments(this.props.match.params.id);
  }

  render() {
    const {
      match: {
        params: { id },
      },
      onDismissError,
    } = this.props;
    const { list } = this.props.historyModel;
    const {
      attachments,
      changeStatusOptionList,
      defaultTexts,
      defaultTextsOptionList,
      error,
      incident,
      loading,
      patching,
      priorityList,
      stadsdeelList,
      statusList,
      typesList,
    } = this.props.incidentModel;
    const { previewState, attachmentHref } = this.state;
    return (
      <Fragment>
        <div className="incident-detail">
          {loading && <LoadingIndicator />}

          {!loading && (
            <Fragment>
              {incident && (
                <Row>
                  <Column span={12}>
                    <DetailHeader
                      incidentId={incident.id}
                      status={incident?.status?.state}
                      links={incident?._links}
                    />
                  </Column>
                </Row>
              )}

              {previewState && (
                <Row>
                  <DetailContainer span={12}>
                    <button
                      aria-label="Sluiten"
                      className="incident-detail__preview-close incident-detail__button--close"
                      type="button"
                      onClick={this.onCloseAll}
                    />

                    {previewState === 'showImage' && (
                      <AttachmentViewer
                        attachments={attachments}
                        href={attachmentHref}
                        onShowAttachment={this.onShowAttachment}
                      />
                    )}

                    {previewState === 'showLocation' && (
                      <LocationPreview location={incident.location} onEditLocation={this.onEditLocation} />
                    )}

                    {previewState === 'editLocation' && (
                      <LocationForm
                        incidentId={incident.id}
                        location={incident.location}
                        onClose={this.onCloseAll}
                      />
                    )}

                    {previewState === 'editStatus' && (
                      <StatusForm
                        incident={incident}
                        patching={patching}
                        error={error}
                        changeStatusOptionList={changeStatusOptionList}
                        defaultTextsOptionList={defaultTextsOptionList}
                        statusList={statusList}
                        defaultTexts={defaultTexts}
                        onDismissError={onDismissError}
                        onClose={this.onCloseAll}
                      />
                    )}
                  </DetailContainer>
                </Row>
              )}

              {!previewState && (
                <Row>
                  <DetailContainer span={7}>
                    {incident && (
                      <Fragment>
                        <Detail
                          incident={incident}
                          attachments={attachments}
                          stadsdeelList={stadsdeelList}
                          onShowLocation={this.onShowLocation}
                          onEditLocation={this.onEditLocation}
                          onShowAttachment={this.onShowAttachment}
                        />

                        <AddNote id={id} />

                        <ChildIncidents incident={incident} />

                        <History list={list} />
                      </Fragment>
                    )}
                  </DetailContainer>

                  <DetailContainer span={4} push={1}>
                    {incident && (
                      <MetaList
                        incident={incident}
                        priorityList={priorityList}
                        typesList={typesList}
                        onEditStatus={this.onEditStatus}
                      />
                    )}
                  </DetailContainer>
                </Row>
              )}
            </Fragment>
          )}
        </div>
      </Fragment>
    );
  }
}

IncidentDetail.defaultProps = {
  previewState: '',
  attachmentHref: '',
};

IncidentDetail.propTypes = {
  previewState: PropTypes.string,
  attachmentHref: PropTypes.string,

  incidentModel: PropTypes.shape({
    incident: incidentType,
    attachments: attachmentsType,
    loading: PropTypes.bool.isRequired,
    patching: PropTypes.shape({
      location: PropTypes.bool,
      status: PropTypes.bool,
    }).isRequired,
    error: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    split: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    stadsdeelList: dataListType,
    priorityList: dataListType,
    typesList: dataListType,
    changeStatusOptionList: dataListType,
    defaultTextsOptionList: dataListType,
    statusList: dataListType,
    defaultTexts: defaultTextsType,
  }).isRequired,
  historyModel: PropTypes.shape({
    list: historyType.isRequired,
  }).isRequired,

  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
  }),

  onRequestIncident: PropTypes.func.isRequired,
  onRequestHistoryList: PropTypes.func.isRequired,
  onRequestAttachments: PropTypes.func.isRequired,
  onRequestDefaultTexts: PropTypes.func.isRequired,
  onDismissError: PropTypes.func.isRequired,
};

/* istanbul ignore next */
const mapStateToProps = () =>
  createStructuredSelector({
    loading: makeSelectLoading(),
    error: makeSelectError(),
    incidentModel: makeSelectIncidentModel,
    historyModel: makeSelectHistoryModel(),
  });

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      onRequestIncident: requestIncident,
      onRequestHistoryList: requestHistoryList,
      onRequestAttachments: requestAttachments,
      onRequestDefaultTexts: requestDefaultTexts,
      onDismissError: dismissError,
    },
    dispatch
  );

export default connect(mapStateToProps, mapDispatchToProps)(IncidentDetail);
