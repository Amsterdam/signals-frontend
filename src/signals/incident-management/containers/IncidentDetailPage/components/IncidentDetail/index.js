import React from 'react';
import PropTypes from 'prop-types';

import { string2date, string2time } from 'shared/services/string-parser/string-parser';
import './style.scss';

class IncidentDetail extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    const { incident } = this.props;
    const extraProperties = incident.extra_properties && Object.keys(incident.extra_properties).map((key) =>
      (<tr key={key}><td>{key}</td><td>{incident.extra_properties[key]}&nbsp;</td></tr>)
    );
    return (
      <div className="incident-detail">
        <div className="incident-detail__body">
          <table className="" cellSpacing="0" cellPadding="0">
            <tbody>
              <tr><td>Datum</td><td>{string2date(incident.created_at)}</td></tr>
              <tr><td>Tijd</td><td>{string2time(incident.created_at)}</td></tr>
              <tr><td>Datum overlast</td><td>{string2date(incident.incident_date_start)}</td></tr>
              <tr><td>Tijd overlast</td><td>{string2time(incident.incident_date_start)}</td></tr>
              <tr><td>Rubriek</td><td>{incident.category.sub}&nbsp;</td></tr>
              <tr><td>Omschrijving</td><td>{incident.text}&nbsp;</td></tr>
              <tr><td>Aanvullende kenmerken</td><td>{incident.text_extra}&nbsp;</td></tr>
              {extraProperties}
              <tr><td>Stadsdeel</td><td>{incident.location.stadsdeel}</td></tr>
              <tr><td>Adres</td><td>{incident.location.address_text}</td></tr>
              <tr><td>Email</td><td>{incident.reporter.email}</td></tr>
              <tr><td>Telefoonnummer</td><td>{incident.reporter.phone}</td></tr>
              <tr><td>Bron</td><td>{incident.source}</td></tr>
              <tr><td>Verantwoordelijke afdeling</td><td>{incident.category.department}&nbsp;</td></tr>
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

IncidentDetail.propTypes = {
  incident: PropTypes.object.isRequired
};

export default IncidentDetail;
