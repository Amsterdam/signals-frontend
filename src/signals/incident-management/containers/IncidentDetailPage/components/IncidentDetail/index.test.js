import React from 'react';
import { shallow } from 'enzyme';

import IncidentDetail from './index';
import stadsdeelList from '../../../../definitions/stadsdeelList';

describe('<IncidentDetail />', () => {
  let incident;

  beforeEach(() => {
    incident = {
      id: 17,
      source: 'Telefoon 14020',
      text: 'snel boot',
      text_extra: '',
      status: { id: 7, text: '', user: null, target_api: '', state: 'm', extern: false, extra_properties: { IP: '83.96.202.117' } },
      location: { id: 5, stadsdeel: 'A', buurt_code: 'A00c', address: { postcode: '1012BL', huisletter: 'B', huisnummer: '84', woonplaats: 'Amsterdam', openbare_ruimte: 'Geldersekade', huisnummer_toevoeging: '' }, address_text: 'Geldersekade 84B- 1012BL Amsterdam', geometrie: { type: 'Point', coordinates: [52.374231891648414, 4.901061058044434] }, extra_properties: null },
      category: { main: 'Overlast op het water', sub: 'Overlast op het water - snel varen', department: '', priority: null },
      reporter: { email: '', phone: '', remove_at: '2018-07-29T07:19:07.863608Z', created_at: '2018-07-15T07:19:08.034968Z', updated_at: '2018-07-15T07:19:08.034982Z', extra_properties: null },
      created_at: '2018-07-15T07:19:08.043593Z',
      incident_date_start: '2018-07-15T07:19:08.043554Z',
      extra_properties: { extra_boten_snelheid_rederij: 'Admiraal Heijn', extra_boten_snelheid_naamboot: 'Speranta', extra_boten_snelheid_rondvaartboot: 'Ja' }
    };
  });

  it('should render correctly', () => {
    const renderedComponent = shallow(
      <IncidentDetail incident={incident} stadsdeelList={stadsdeelList} />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render correctly without extra properties', () => {
    delete incident.extra_properties;
    const renderedComponent = shallow(
      <IncidentDetail incident={incident} stadsdeelList={stadsdeelList} />
    );
    expect(renderedComponent.find('tr').length).toEqual(13);
    expect(renderedComponent).toMatchSnapshot();
  });
});
