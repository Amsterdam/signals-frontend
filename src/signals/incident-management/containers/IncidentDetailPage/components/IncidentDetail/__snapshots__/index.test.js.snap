// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncidentDetail /> should render correctly 1`] = `
<div
  className="incident-detail"
>
  <div
    className="incident-detail__body"
  >
    <table
      cellPadding="0"
      cellSpacing="0"
      className=""
    >
      <tbody>
        <tr>
          <td>
            Datum
          </td>
          <td>
            15-07-2018
          </td>
        </tr>
        <tr>
          <td>
            Tijd
          </td>
          <td>
            09:19
          </td>
        </tr>
        <tr>
          <td>
            Datum overlast
          </td>
          <td>
            15-07-2018
          </td>
        </tr>
        <tr>
          <td>
            Tijd overlast
          </td>
          <td>
            09:19
          </td>
        </tr>
        <tr>
          <td>
            Rubriek
          </td>
          <td>
            Overlast op het water - snel varen
             
          </td>
        </tr>
        <tr>
          <td>
            Onschrijving
          </td>
          <td>
            snel boot
             
          </td>
        </tr>
        <tr>
          <td>
            Aanvullende kenmerken
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          key="extra_boten_snelheid_rederij"
        >
          <td>
            extra_boten_snelheid_rederij
          </td>
          <td>
            Admiraal Heijn
             
          </td>
        </tr>
        <tr
          key="extra_boten_snelheid_naamboot"
        >
          <td>
            extra_boten_snelheid_naamboot
          </td>
          <td>
            Speranta
             
          </td>
        </tr>
        <tr
          key="extra_boten_snelheid_rondvaartboot"
        >
          <td>
            extra_boten_snelheid_rondvaartboot
          </td>
          <td>
            Ja
             
          </td>
        </tr>
        <tr>
          <td>
            Stadsdeel
          </td>
          <td>
            A
          </td>
        </tr>
        <tr>
          <td>
            Adres
          </td>
          <td>
            Geldersekade 84B- 1012BL Amsterdam
          </td>
        </tr>
        <tr>
          <td>
            Email
          </td>
          <td />
        </tr>
        <tr>
          <td>
            Telefoonnummer
          </td>
          <td />
        </tr>
        <tr>
          <td>
            Bron
          </td>
          <td>
            Telefoon 14020
          </td>
        </tr>
        <tr>
          <td>
            Verantwoordelijke afdeling
          </td>
          <td>
             
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<IncidentDetail /> should render correctly without extra properties 1`] = `
<div
  className="incident-detail"
>
  <div
    className="incident-detail__body"
  >
    <table
      cellPadding="0"
      cellSpacing="0"
      className=""
    >
      <tbody>
        <tr>
          <td>
            Datum
          </td>
          <td>
            15-07-2018
          </td>
        </tr>
        <tr>
          <td>
            Tijd
          </td>
          <td>
            09:19
          </td>
        </tr>
        <tr>
          <td>
            Datum overlast
          </td>
          <td>
            15-07-2018
          </td>
        </tr>
        <tr>
          <td>
            Tijd overlast
          </td>
          <td>
            09:19
          </td>
        </tr>
        <tr>
          <td>
            Rubriek
          </td>
          <td>
            Overlast op het water - snel varen
             
          </td>
        </tr>
        <tr>
          <td>
            Onschrijving
          </td>
          <td>
            snel boot
             
          </td>
        </tr>
        <tr>
          <td>
            Aanvullende kenmerken
          </td>
          <td>
             
          </td>
        </tr>
        <tr>
          <td>
            Stadsdeel
          </td>
          <td>
            A
          </td>
        </tr>
        <tr>
          <td>
            Adres
          </td>
          <td>
            Geldersekade 84B- 1012BL Amsterdam
          </td>
        </tr>
        <tr>
          <td>
            Email
          </td>
          <td />
        </tr>
        <tr>
          <td>
            Telefoonnummer
          </td>
          <td />
        </tr>
        <tr>
          <td>
            Bron
          </td>
          <td>
            Telefoon 14020
          </td>
        </tr>
        <tr>
          <td>
            Verantwoordelijke afdeling
          </td>
          <td>
             
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
